kernel/aster-nix/src/sched/priority_scheduler.rs

```diff
@@ -17,40 +22,199 @@ pub fn init() {
 /// are always prioritized during scheduling.
 /// Normal tasks are placed in the `normal_tasks` queue and are only
 /// scheduled for execution when there are no real-time tasks.
-struct PreemptScheduler {
+struct PreemptScheduler<T: PreemptSchedInfo> {
+    rq: Vec<SpinLock<PreemptRunQueue<T>>>,
+}
+
+impl<T: PreemptSchedInfo> PreemptScheduler<T> {
+    fn new(ncpu: u32) -> Self {
+        let mut rq = Vec::with_capacity(ncpu as usize);
+        for _ in 0..ncpu {
+            rq.push(SpinLock::new(PreemptRunQueue::new()));
+        }
+        Self { rq }
+    }
+
+    /// Selects a cpu for task to run on.
+    fn select_cpu(&self, _runnable: &Arc<T>) -> usize {
+        // FIXME: adopt more reasonable policy once we fully enable SMP.
+        0
+    }
+}
+
+impl<T: Sync + Send + PreemptSchedInfo> Scheduler<T> for PreemptScheduler<T> {
+    fn enqueue(&self, runnable: Arc<T>, flags: EnqueueFlags) -> Option<u32> {
+        if flags == EnqueueFlags::Wake {
+            for cpu_id in 0..self.rq.len() {
+                let rq = self.rq[cpu_id].lock_irq_disabled();
+                if rq.current.is_none() {
+                    continue;
+                }
+                if Arc::ptr_eq(&runnable, &rq.current.as_ref().unwrap().runnable) {
```

Comment:
Visiting the runqueue for every CPU  is not an efficient way to decide if `runnable` is the current running task on any CPU. Checking a field of `Task` is more efficient. Is there any technical difficulty to do so?

In addition, we have talked about it is one of the responsibilities of the scheduler to prevent a task from being enqueued multiple times in a single or multiple runqueues. This can also be achieved by adding a field to `Task`.

Reply:
I didn't find out a elegant solution with one additional field of `Task` at that time so I added this expensive check. However, I later found this check was not even correct.

I posted a correct but maybe unsatisfying solution in the latest commit. It's based on an additional field in `Task` called `cpu`, which indicates which CPU the task is on. I chosed an `Option<u32>` instead of a `bool` to ensure that we can lock the same runqueue in `unpark_current -> enqueue` as we do in `park_current -> reschedule` therefore ensure the desired concurrency safety.